File: src/game/solver.ts
/**
 * Solver for the color-sort puzzle. Used to (a) verify that a generated level is
 * actually solvable — the hard guarantee behind the generator — and (b) compute a
 * known solution and its step count. The same machinery can later power a "hint" tool.
 */
import { isWon, legalMoves, pour, topColor } from './engine';
import { funnelAccepts, type FunnelGrid } from './funnels';
import { stateKey } from './search';
import type { GameState, Move } from './types';

/**
 * An order-independent string key for a (full-information) state — the no-concealment case of
 * the shared {@link stateKey}. Two boards that differ only by bottle ordering collapse to the
 * same key, which dramatically shrinks the search space.
 */
export function canonical(state: GameState): string {
  return stateKey(state);
}

interface SolveOptions {
  /** Safety cap on explored nodes so generation can never hang. Default 200k. */
  maxNodes?: number;
  /**
   * Color-locked funnels overlay (chapter 2+). When present, a pour into a tube whose tint rejects
   * the poured color is omitted from the search — the same constraint the player plays under. Default
   * `undefined` ⇒ no funnels, so every search behaves exactly as the un-funneled game.
   */
  funnels?: FunnelGrid;
}

/**
 * Decide whether a candidate pour is worth exploring. Prunes moves that can never help:
 * - pouring a bottle that is already a finished single color, and
 * - pouring into one of several interchangeable empty bottles (only the first matters).
 *
 * With `funnels`, a pour the destination tube's tint rejects is also pruned (it isn't a legal player
 * move). Folding the funnel gate in here means every search built on `isUsefulMove` — `search`,
 * `isStuckInLoop`, `bfsOptimal`, and `usefulMoves` (the difficulty metrics) — honors funnels through
 * one shared check.
 */
function isUsefulMove(state: GameState, from: number, to: number, funnels?: FunnelGrid): boolean {
  const src = state.bottles[from]!;
  const dst = state.bottles[to]!;

  const color = topColor(src)!;
  // A funnel rejects any inflow that isn't its tint — not a legal move to consider.
  if (!funnelAccepts(funnels, to, color)) return false;

  // Don't disturb a bottle that is already uniformly one color filling it entirely
  // unless the destination is also that same color (consolidation is pointless here).
  const srcUniform = src.length > 0 && src.every((c) => c === src[0]);
  if (srcUniform && dst.length === 0) return false; // moving a solid block to empty = no progress

  // Among equivalent empty destinations, only consider the first one — but funnels make empties
  // non-interchangeable, so the representative is the first empty that ALSO accepts this color (an
  // empty locked to another color is a different destination, not a redundant copy of this one).
  if (dst.length === 0) {
    const firstEmpty = state.bottles.findIndex(
      (b, idx) => b.length === 0 && funnelAccepts(funnels, idx, color),
    );
    if (to !== firstEmpty) return false;
  }
  return true;
}

/**
 * The legal moves worth considering from a state — `legalMoves` minus the ones `isUsefulMove`
 * prunes (disturbing a finished solid block, pouring into a redundant empty, or — under `funnels` —
 * a pour a destination funnel rejects). This is the branching the search actually explores, so
 * difficulty metrics (branching factor, forced-move ratio in `difficulty.ts`) measure against it
 * rather than the raw legal-move count.
 */
export function usefulMoves(
  state: GameState,
  funnels?: FunnelGrid,
): Array<{ from: number; to: number }> {
  return legalMoves(state).filter(({ from, to }) => isUsefulMove(state, from, to, funnels));
}

export interface SolveResult {
  /** A winning move sequence, or `null` if none was found. */
  solution: Move[] | null;
  /**
   * True when the search explored the entire reachable state space without hitting the
   * node budget. A `null` solution with `exhausted: true` is a *proof* that the board is
   * unsolvable; a `null` solution with `exhausted: false` is merely inconclusive (the
   * search gave up early), so callers must not treat it as a deadlock.
   */
  exhausted: boolean;
}

/**
 * Depth-first search for *a* solution, reporting whether the search was exhaustive.
 * This is the primitive behind `solve`, `isSolvable`, and `isUnsolvable`.
 */
export function search(state: GameState, options: SolveOptions = {}): SolveResult {
  const maxNodes = options.maxNodes ?? 200_000;
  const visited = new Set<string>();
  let nodes = 0;
  let hitCap = false;

  const dfs = (current: GameState, path: Move[]): Move[] | null => {
    if (isWon(current)) return path;
    if (++nodes > maxNodes) {
      hitCap = true;
      return null;
    }

    const key = canonical(current);
    if (visited.has(key)) return null;
    visited.add(key);

    for (const { from, to } of legalMoves(current)) {
      if (!isUsefulMove(current, from, to, options.funnels)) continue;
      const { state: next, move } = pour(current, from, to);
      const result = dfs(next, [...path, move]);
      if (result) return result;
    }
    return null;
  };

  const solution = dfs(state, []);
  return { solution, exhausted: !hitCap };
}

/**
 * Find a solution, or `null` if the board is unsolvable or the node budget is exhausted
 * (both treated as "reject" by the generator).
 */
export function solve(state: GameState, options?: SolveOptions): Move[] | null {
  return search(state, options).solution;
}

/** Whether a state can be solved at all (thin wrapper over `search`). */
export function isSolvable(state: GameState, options?: SolveOptions): boolean {
  return search(state, options).solution !== null;
}

/**
 * Proof-based deadlock: the board is unwinnable and the solver has *exhaustively* shown
 * it. Unlike `engine.isDeadlocked` (which only catches boards with zero legal moves),
 * this also detects "loops" — boards where moves remain but none can ever lead to a win.
 * Returns `false` if the search was inconclusive (hit the node budget), so a winnable
 * game is never ended by mistake.
 */
export function isUnsolvable(state: GameState, options?: SolveOptions): boolean {
  const { solution, exhausted } = search(state, options);
  return solution === null && exhausted;
}

/**
 * Whether the player is provably going in circles: every state reachable from `state` (under the
 * same useful-move pruning the solver explores) has *already been visited this attempt* — so no
 * move can ever produce a board the player hasn't already seen; the only moves left loop back.
 *
 * This is the trigger for the gentle "going in circles" nudge, distinct from `engine.isDeadlocked`
 * (a hard zero-moves wall) and from `isUnsolvable` (which would fire the instant a board becomes
 * unwinnable — even with plenty of fresh states still to explore, which we deliberately don't
 * interrupt). Reusing `isUsefulMove` is what makes the distinction meaningful: parking a finished
 * solid block in an empty tube is not a real "somewhere new to go", so a spare empty bottle can't
 * disguise a dead end.
 *
 * Conservative by construction: it returns `false` the moment it finds any unvisited reachable
 * state, and also `false` if the search hits the node budget (inconclusive) — so a player who
 * still has somewhere meaningful to go is never nagged.
 */
export function isStuckInLoop(
  state: GameState,
  visited: ReadonlySet<string>,
  options: SolveOptions = {},
): boolean {
  const maxNodes = options.maxNodes ?? 20_000;
  const seen = new Set<string>([canonical(state)]);
  const stack: GameState[] = [state];
  let nodes = 0;

  while (stack.length > 0) {
    const current = stack.pop()!;
    if (++nodes > maxNodes) return false; // inconclusive — don't nag
    for (const { from, to } of legalMoves(current)) {
      if (!isUsefulMove(current, from, to, options.funnels)) continue;
      const { state: next } = pour(current, from, to);
      if (isWon(next)) return false; // a win is reachable (and would be unvisited anyway)
      const key = canonical(next);
      if (!visited.has(key)) return false; // somewhere new to go → not circling
      if (seen.has(key)) continue;
      seen.add(key);
      stack.push(next);
    }
  }
  return true; // whole reachable closure already visited — only loops remain
}

/**
 * Breadth-first search for the *minimum* number of moves. Heavier than `solve`, so it's
 * intended only for difficulty calibration on small boards. Returns `null` if unsolved
 * within the node budget.
 */
export function bfsOptimal(state: GameState, options: SolveOptions = {}): number | null {
  const maxNodes = options.maxNodes ?? 200_000;
  if (isWon(state)) return 0;

  const visited = new Set<string>([canonical(state)]);
  let frontier: GameState[] = [state];
  let depth = 0;
  let nodes = 0;

  while (frontier.length > 0) {
    const next: GameState[] = [];
    depth++;
    for (const current of frontier) {
      for (const { from, to } of legalMoves(current)) {
        if (!isUsefulMove(current, from, to, options.funnels)) continue;
        const { state: child } = pour(current, from, to);
        if (isWon(child)) return depth;
        const key = canonical(child);
        if (visited.has(key)) continue;
        visited.add(key);
        if (++nodes > maxNodes) return null;
        next.push(child);
      }
    }
    frontier = next;
  }
  return null;
}

File: core/src/solver.rs
//! Full-information solver — port of `src/game/solver.ts`. The DFS explores the identical
//! node sequence as JS (same move enumeration order, same prune order, same node-budget
//! placement), so it returns the identical first solution — the property the bake's
//! byte-reproducibility rests on. The solver consults only `funnels` (deliberately not
//! ice-aware, same as JS).

use std::collections::HashSet;

use crate::engine::{is_won, legal_moves, pour};
use crate::funnels::{accepts, Funnels};
use crate::state::{state_key, Key, State};
use crate::types::Move;

pub const DEFAULT_MAX_NODES: usize = 200_000;

/// Order-independent canonical key for a full-information state.
pub fn canonical(state: &State) -> Key {
    state_key(state, None)
}

/// Prune moves that can never help — port of `isUsefulMove` (finished-solid-block moves,
/// redundant empties collapsed to the first ACCEPTING empty, funnel-rejected pours).
fn is_useful_move(state: &State, from: usize, to: usize, funnels: Option<&Funnels>) -> bool {
    let src = &state.tubes[from];
    let dst = &state.tubes[to];

    let color = src.top().unwrap();
    if !accepts(funnels, to, color) {
        return false;
    }

    let src_uniform = src.is_uniform();
    if src_uniform && dst.is_empty() {
        return false; // moving a solid block to empty = no progress
    }

    if dst.is_empty() {
        let first_empty = state
            .tubes
            .iter()
            .enumerate()
            .position(|(idx, b)| b.is_empty() && accepts(funnels, idx, color));
        if Some(to) != first_empty {
            return false;
        }
    }
    true
}

/// The legal moves worth considering — `legal_moves` minus the `is_useful_move` prunes, in
/// canonical order.
pub fn useful_moves(state: &State, funnels: Option<&Funnels>) -> Vec<(u8, u8)> {
    legal_moves(state)
        .into_iter()
        .filter(|&(from, to)| is_useful_move(state, from as usize, to as usize, funnels))
        .collect()
}

pub struct SolveResult {
    pub solution: Option<Vec<Move>>,
    /// True when the whole reachable space was explored within budget — a `None` solution
    /// with `exhausted` is a *proof* of unsolvability.
    pub exhausted: bool,
}

struct Frame {
    state: State,
    moves: Vec<(u8, u8)>,
    next: usize,
}

/// Iterative pre-order DFS for *a* solution — structurally identical to the JS `search`
/// (goal check before the node count, node count before the visited check), so node budgets
/// bite at the same points and the first solution found is the same.
pub fn search(state: &State, funnels: Option<&Funnels>, max_nodes: usize) -> SolveResult {
    let mut visited: HashSet<Key> = HashSet::new();
    let mut nodes = 0usize;
    let mut hit_cap = false;

    let mut path: Vec<Move> = Vec::new();
    let mut stack: Vec<Frame> = Vec::new();

    enum Entered {
        Won,
        Pruned,
        Pushed,
    }

    let enter = |current: State,
                 stack: &mut Vec<Frame>,
                 nodes: &mut usize,
                 hit_cap: &mut bool,
                 visited: &mut HashSet<Key>|
     -> Entered {
        if is_won(&current) {
            return Entered::Won;
        }
        *nodes += 1;
        if *nodes > max_nodes {
            *hit_cap = true;
            return Entered::Pruned;
        }
        let key = canonical(&current);
        if !visited.insert(key) {
            return Entered::Pruned;
        }
        let moves = useful_moves(&current, funnels);
        stack.push(Frame {
            state: current,
            moves,
            next: 0,
        });
        Entered::Pushed
    };

    if matches!(
        enter(
            state.clone(),
            &mut stack,
            &mut nodes,
            &mut hit_cap,
            &mut visited
        ),
        Entered::Won
    ) {
        return SolveResult {
            solution: Some(Vec::new()),
            exhausted: !hit_cap,
        };
    }

    while let Some(frame) = stack.last_mut() {
        if frame.next >= frame.moves.len() {
            stack.pop();
            if !stack.is_empty() {
                path.pop(); // a non-root frame: drop the move that led into it
            }
            continue;
        }
        let (from, to) = frame.moves[frame.next];
        frame.next += 1;
        let (next, mv) = pour(&frame.state, from as usize, to as usize, usize::MAX);
        path.push(mv);
        match enter(next, &mut stack, &mut nodes, &mut hit_cap, &mut visited) {
            Entered::Won => {
                return SolveResult {
                    solution: Some(path),
                    exhausted: !hit_cap,
                }
            }
            Entered::Pruned => {
                path.pop();
            }
            Entered::Pushed => {}
        }
    }

    SolveResult {
        solution: None,
        exhausted: !hit_cap,
    }
}

pub fn solve(state: &State, funnels: Option<&Funnels>, max_nodes: usize) -> Option<Vec<Move>> {
    search(state, funnels, max_nodes).solution
}

pub fn is_solvable(state: &State, funnels: Option<&Funnels>, max_nodes: usize) -> bool {
    solve(state, funnels, max_nodes).is_some()
}

pub fn is_unsolvable(state: &State, funnels: Option<&Funnels>, max_nodes: usize) -> bool {
    let r = search(state, funnels, max_nodes);
    r.solution.is_none() && r.exhausted
}

/// Whether every reachable state has already been visited this attempt — port of
/// `isStuckInLoop` (conservative: inconclusive budget or any fresh state ⇒ `false`).
pub fn is_stuck_in_loop(
    state: &State,
    visited: &HashSet<Key>,
    funnels: Option<&Funnels>,
    max_nodes: usize,
) -> bool {
    let mut seen: HashSet<Key> = HashSet::from([canonical(state)]);
    let mut stack: Vec<State> = vec![state.clone()];
    let mut nodes = 0usize;

    while let Some(current) = stack.pop() {
        nodes += 1;
        if nodes > max_nodes {
            return false; // inconclusive — don't nag
        }
        for (from, to) in legal_moves(&current) {
            if !is_useful_move(&current, from as usize, to as usize, funnels) {
                continue;
            }
            let (next, _) = pour(&current, from as usize, to as usize, usize::MAX);
            if is_won(&next) {
                return false;
            }
            let key = canonical(&next);
            if !visited.contains(&key) {
                return false; // somewhere new to go
            }
            if !seen.insert(key) {
                continue;
            }
            stack.push(next);
        }
    }
    true
}

/// Layered BFS for the minimum full-information move count — port of `bfsOptimal` (node
/// budget counted on NEW states, after the visited check, exactly like JS).
pub fn bfs_optimal(state: &State, funnels: Option<&Funnels>, max_nodes: usize) -> Option<usize> {
    if is_won(state) {
        return Some(0);
    }
    let mut visited: HashSet<Key> = HashSet::from([canonical(state)]);
    let mut frontier: Vec<State> = vec![state.clone()];
    let mut depth = 0usize;
    let mut nodes = 0usize;

    while !frontier.is_empty() {
        let mut next_frontier: Vec<State> = Vec::new();
        depth += 1;
        for current in &frontier {
            for (from, to) in legal_moves(current) {
                if !is_useful_move(current, from as usize, to as usize, funnels) {
                    continue;
                }
                let (child, _) = pour(current, from as usize, to as usize, usize::MAX);
                if is_won(&child) {
                    return Some(depth);
                }
                let key = canonical(&child);
                if !visited.insert(key) {
                    continue;
                }
                nodes += 1;
                if nodes > max_nodes {
                    return None;
                }
                next_frontier.push(child);
            }
        }
        frontier = next_frontier;
    }
    None
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::state::Tube;

    fn state(tubes: Vec<&[u8]>, capacity: u8) -> State {
        State {
            tubes: tubes.into_iter().map(Tube::from_cells).collect(),
            capacity,
        }
    }

    #[test]
    fn solves_a_trivial_board() {
        let s = state(vec![&[1, 1, 1], &[1], &[]], 4);
        let solution = solve(&s, None, DEFAULT_MAX_NODES).expect("solvable");
        assert!(!solution.is_empty());
        let mut cur = s;
        for m in &solution {
            cur = pour(&cur, m.from as usize, m.to as usize, usize::MAX).0;
        }
        assert!(is_won(&cur));
    }

    #[test]
    fn proves_unsolvable_when_exhausted() {
        // Two full mixed tubes, no free space: exhaustively unwinnable.
        let s = state(vec![&[0, 1, 0, 1], &[1, 0, 1, 0]], 4);
        assert!(is_unsolvable(&s, None, DEFAULT_MAX_NODES));
    }

    #[test]
    fn bfs_matches_obvious_optimum() {
        let s = state(vec![&[1, 1, 1], &[1], &[]], 4);
        assert_eq!(bfs_optimal(&s, None, DEFAULT_MAX_NODES), Some(1));
    }
}
